### PR TITLE
fix: detach spikes recorded by STDP learners to prevent memory growth

### LINKED ISSUE #576: STDPLearner never frees the created tensors even when out of scope, causing torch.OutOfMemoryError
**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**SpikingJelly version**

`0.0.0.0.15`

**Description**

I am performing an hyperparameter search, using a grid search, which implies creating a different model every loop and training it. The instantiation and training of the model are performed inside a function that is called with different configuration parameters every run.

When finishing every iteration of the loop, the model and all created variables, including optimizer, dataset, and STDPLearners, go out of scope, which should delete them. However, the current behaviour is that tensors created by STDPLearner are never deleted, filling up the memory until the program crashes. The MWE crashes after only 3 iterations when the model includes a `layer.Conv2D`, and the problem still exists when the model only uses `layer.Linear` layers, although to a lesser extent.

Happens with both torch and cupy backends. Explicitly calling the garbage collector does not help. When training using gradient descent, as shown in the minimal example by swapping `train_model_stdp` with the `train_model_gd` function, this problem doesn't show up, linking it with the `STDPLearner` class.

**Minimal code to reproduce the error/bug**

This is the basic training code that causes the issue. A single epoch per run is used because the problem only shows up when starting new runs. Includes a debugging function to see how many tensors are there.

```python
import itertools
import gc

import torch
import torch.nn as nn
import torchvision
import torch.utils.data as D
from spikingjelly.activation_based import layer, neuron, functional, learning
from tqdm.auto import tqdm


def f_weight(x):
    return torch.clamp(x, -1., 1.)


class ConvMNIST(nn.Module):
    def __init__(self, tau: float):
        super().__init__()

        self.network = nn.Sequential(
            layer.Conv2d(1, 32, kernel_size=3, stride=1, bias=False),
            neuron.IFNode(),
            layer.AvgPool2d(2, 2),

            layer.Flatten(),
            layer.Linear(13*13*32, 10, bias=False),
            neuron.LIFNode(tau=tau)
        )

        functional.set_step_mode(self.network, 'm')
        functional.set_backend(self.network, 'torch')

    def forward(self,x):
        return self.network(x)
    
    def get_stdp_learners(self):
        stdp_learners = []

        for i in range(len(self.network)):
            if isinstance(self.network[i], (layer.Conv2d, layer.Linear)):
                stdp_learners.append(
                    learning.STDPLearner(step_mode='m', synapse=self.network[i], sn=self.network[i+1], tau_pre=2.0, tau_post=2.0,
                                        f_pre=f_weight, f_post=f_weight)
                )
        
        return stdp_learners


def train_model_stdp(
        model: nn.Module,
        learning_rate: float,
        train_data: D.Dataset,
        device="cuda"):

    model.to(device)

    train_data = D.Subset(train_data, range(400))
    train_data = D.DataLoader(train_data, batch_size=64, shuffle=True, drop_last=True, num_workers=4)

    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.0)

    stdp_learners = model.get_stdp_learners()

    print_tensors()
    for epoch in range(1, 2):
        model.train()
        for l in stdp_learners:
            l.enable()

        for batch, targets in (
            pbar := tqdm(train_data, dynamic_ncols=True, desc=f"Epoch {epoch}")
        ):

            batch = batch.to(device)
            targets = targets.to(device)

            output = model(batch.expand(100, -1, -1, -1, -1))

            optimizer.zero_grad()
            for l in stdp_learners:
                l.step()
            optimizer.step()

            functional.reset_net(model)
            for l in stdp_learners:
                l.reset()
    
    print_tensors()

def print_tensors(show_tensors=False):
    tensor_count = 0
    for obj in gc.get_objects():
        try:
            if torch.is_tensor(obj) or (hasattr(obj, 'data') and torch.is_tensor(obj.data)):
                if show_tensors: print(type(obj), obj.size())
                tensor_count += 1
        except: pass
    print(f"{tensor_count=}")

def run(tau: float, learning_rate: float):
    model = ConvMNIST(tau)

    train_dataset = torchvision.datasets.MNIST(
        root='./datasets',
        train=True,
        transform=torchvision.transforms.ToTensor(),
        download=True
    )

    train_model_stdp(model=model, learning_rate=learning_rate, train_data=train_dataset)

if __name__ == '__main__':
    search_space = {
        "tau": [2.0, 2.5, 3.0, 5.0, 10.0, 100.0],
        "learning_rate": [0.1, 0.01, 1e-3, 1e-4, 1e-5]
    }

    n_runs = 0
    for tau, learning_rate in itertools.product(*search_space.values()):
        n_runs += 1
        print(f"Starting run {n_runs} with params: {tau=} {learning_rate=}")
        run(tau=tau, learning_rate=learning_rate)
        print(gc.collect())
        print_tensors()
```

By swapping the `train_model_stdp` function with the following one, which simulates training with gradient descent, the problem disappears

```python
def train_model_gd(
        model: nn.Module,
        learning_rate: float,
        train_data: D.Dataset,
        device="cuda"):

    model.to(device)

    train_data = D.Subset(train_data, range(400))
    train_data = D.DataLoader(train_data, batch_size=64, shuffle=True, drop_last=True, num_workers=4)

    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.0)

    print_tensors()
    for epoch in range(1, 2):
        model.train()

        for batch, targets in (
            pbar := tqdm(train_data, dynamic_ncols=True, desc=f"Epoch {epoch}")
        ):

            batch = batch.to(device)
            targets = targets.to(device)

            output = model(batch.expand(100, -1, -1, -1, -1))
            mean_activations = output.mean(0)
            batch_loss = nn.functional.cross_entropy(mean_activations, targets)

            optimizer.zero_grad()
            batch_loss.backward()
            optimizer.step()

            functional.reset_net(model)
    
    print_tensors()
```

### PR DESCRIPTION
**Pull request by JPL11**
## What

Detach the spike tensors recorded by the monitors inside `STDPLearner`, `MSTDPLearner`, and `MSTDPETLearner`, so the learners never build or retain the autograd graph of the network's forward pass. Adds `test/activation_based/test_learning.py` with a leak regression test and STDP-correctness tests, plus a CHANGELOG entry.

Fixes #576.

## Why

`STDPLearner`'s `InputMonitor`/`OutputMonitor` record graph-connected spike tensors. When `step()` runs with grad mode enabled (the natural way to call it — the issue's MWE and the STDP tutorial both do), three things go wrong:

1. `trace_pre`/`trace_post` are computed from those spikes, so the traces accumulate an ever-growing autograd graph across batches (traces are recurrent).
2. `step(on_grad=True)` writes a **graph-connected** tensor into `synapse.weight.grad`, keeping the whole forward graph alive through the parameter.
3. The retained graphs survive `gc.collect()` and accumulate every time a model + learner is re-created, e.g. in a hyperparameter search — the OOM reported in #576.

Measured on the issue's scenario (5 model re-creations, CPU, torch 2.13): **before** — 414 leaked tensors / 18.8M elements / ~70 MB RSS growth *per run*, unbounded; **after** — 0 leaked tensors, flat RSS.

## Why detach-at-record rather than `torch.no_grad()` in `step()`

Wrapping `step()` in `no_grad` (the workaround recommended in #576) stops graphs from being built during the update, but the recorded spikes still pin each forward pass's graph between `forward()` and `step()` — significant transient memory in pure-STDP training where no `backward()` ever frees it. Detaching at record time is strictly stronger: no graph is ever built or retained by the learners, and users no longer need the `no_grad` wrapper at all.

Gradient flow *through* STDP updates was never meaningfully supported: `f_pre`/`f_post` already operate on `weight.data`, and the recommended workaround in #576 already excluded `step()` from the graph. If some future use case needs differentiable `delta_w`, that would be a deliberate feature, not this bug.

## Verification

- New regression test `test_stdp_learner_frees_tensors_across_runs` fails on master (leaked tensors accumulate) and passes with this fix; `test_stdp_learner_matches_functional_update` confirms `delta_w` is numerically unchanged.
- Issue-style MWE (Conv2d + Linear, multi-step, grad-mode `step()`, model re-created per run): RSS flat across 5 runs, 0 tensors surviving gc (was +414 tensors / +70 MB per run).
- Full suite: 1326 passed, 144 skipped (CPU) — no regressions.
- `uv format` clean.

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed potential unbounded memory growth and out-of-memory errors when repeatedly recreating STDP learners in loops (e.g., hyperparameter search).
  - Learner monitor recordings (spikes/monitor data) are now detached from the forward autograd graph.
  - `step()` no longer needs to be wrapped in `torch.no_grad()` for this memory-safety behavior.

- **Tests**
  - Added regression tests covering monitor detachment, graph retention behavior, weight-update consistency across variants, and tensor cleanup across repeated runs.

### CODE REVIEW

File: test/activation_based/test_learning.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+            functional.reset_net(net)
+            for l in learners:
+                l.reset()
+
+    one_run()
+    baseline = _count_alive_tensors()
+    for _ in range(3):
+        one_run()
+    assert _count_alive_tensors() == baseline
```
[the remark this answers is not in the document]
Comment (JPL11) [reply test/activation_based/test_learning.py]: Good point — applied in ea36163. A real leak accumulates hundreds of tensors per run here, so the tolerance doesn't weaken the signal.

File: test/activation_based/test_learning.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+def test_stdp_learner_records_are_detached():
+    net = _build_net("m")
+    learners = _build_learners(net, "m")
+
+    x = (torch.rand(4, 2, 1, 8, 8) > 0.5).float()
+    y = net(x)
+
+    assert y.requires_grad
+    for l in learners:
```
[the remark this answers is not in the document]
Comment (JPL11) [reply test/activation_based/test_learning.py]: Renamed throughout in ea36163, and also prefixed the unused unpacked traces with underscores.

**Comment by AllenYolk** [comment]
Thanks for the fix.

The root cause matches #576: the learner monitors were keeping graph-connected spike tensors, which then propagated into traces / `weight.grad` and retained the forward graph across runs. Detaching at record time is the right place to fix this, since it prevents the learner from retaining the graph before `step()` is even called.

Given that STDP updates already operate through `weight.data`, and the previous workaround was to call `step()` under `torch.no_grad()`, this does not appear to remove a supported differentiable-STDP path. LGTM.
